crates/matrix-sdk-crypto/src/verification/requests.rs

```diff
@@ -1046,283 +1152,425 @@ struct Ready {
-                            user_id = self.other_user_id.as_str(),
-                            device_id = self.state.other_device_id.as_str(),
-                            "Can't create a QR code, the user's identity \
-                             doesn't have a valid master key"
+                            user_id = request_state.other_user_id.as_str(),
+                            device_id = state.other_device_id.as_str(),
+                            "Can't create a QR code, we don't trust our own \
+                                 master key"
                         );
                         None
                     }
+                } else {
+                    warn!(
+                        user_id = request_state.other_user_id.as_str(),
+                        device_id = state.other_device_id.as_str(),
+                        "Can't create a QR code, the user's identity \
+                             doesn't have a valid master key"
+                    );
+                    None
                 }
             }
-        } else {
-            warn!(
-                user_id = self.other_user_id.as_str(),
-                device_id = self.state.other_device_id.as_str(),
-                "Can't create a QR code, the user doesn't have a valid cross \
+        }
+    } else {
+        warn!(
+            user_id = request_state.other_user_id.as_str(),
+            device_id = state.other_device_id.as_str(),
+            "Can't create a QR code, the user doesn't have a valid cross \
                  signing identity."
-            );
+        );
 
-            None
+        None
+    };
+
+    if let Some(verification) = verification {
+        let new_state = RequestState {
+            verification_cache: request_state.verification_cache.to_owned(),
+            store: request_state.store.to_owned(),
+            flow_id: request_state.flow_id.to_owned(),
+            other_user_id: request_state.other_user_id.to_owned(),
+            state: Transitioned {
+                ready: state.to_owned(),
+                verification: verification.to_owned().into(),
+            },
         };
 
-        if let Some(verification) = &verification {
-            self.verification_cache.insert_qr(verification.clone());
-        }
+        request_state.verification_cache.insert_qr(verification.to_owned());
 
-        Ok(verification)
+        Ok(Some((new_state, verification)))
+    } else {
+        Ok(None)
     }
+}
 
-    async fn receive_start(
-        &self,
-        sender: &UserId,
-        content: &StartContent<'_>,
-        we_started: bool,
-        request_handle: RequestHandle,
-    ) -> Result<(), CryptoStoreError> {
-        info!(
-            sender = sender.as_str(),
-            device = content.from_device().as_str(),
-            "Received a new verification start event",
-        );
-
-        let Some(device) = self.store.get_device(sender, content.from_device()).await? else {
+async fn foo<T: Clone>(
```

Comment:
Is `foo` the final name?

Reply:
I can change the name, but I can't promise that it'll be better than `foo`.